backend/aws_helpers/dynamo_db_utils/dynamo_db_utils.py

```diff
@@ -0,0 +1,212 @@
+        partition_key_value (str): Value of partition key to match
+
+    Returns:
+        dict: One item from DynamoDB table in JSON format
+
+    Raises:
+        ValueError: If table_name is not a valid table name
+        Exception: If DynamoDB get_item call fails for any reason, e.g., item not found
+    """
+    # Validation steps
+    if table_name not in ALL_DYANMODB_TABLES.keys():
+        raise ValueError("Invalid table name: " + table_name)
+
+    # Get item
+    partition_key = ALL_DYANMODB_TABLES[table_name]["partition_key"]
+    item_key = {partition_key: partition_key_value}
+    table = dynamodb.Table(table_name)
+    response = table.get_item(Key=item_key)
+    if response.get("Item") is None:
+        raise Exception("Item not found")
+
+    return response["Item"]
+
+
+def get_dynamo_items_by_gsi(table_name: str, gsi_value: Union[int, str]) -> list[dict]:
+    """
+    Get items from DynamoDB table that match the given GSI value, returns [] if no items match
+
+    Args:
+        table_name (str): Name of DynamoDB table
+        gsi_value (int or str): Value of GSI to match
+
+    Returns:
+        list[dict]: One array containing all matching items from DynamoDB table in JSON format
+
+    Raises:
+        ValueError: If table_name is not a valid table name
+        Exception: If DynamoDB get_item call fails for any reason
+    """
+    # Validation steps
+    if table_name not in ALL_DYANMODB_TABLES.keys():
+        raise ValueError("Invalid table name: " + table_name)
+    gsi_key = ALL_DYANMODB_TABLES[table_name].get("gsi")
+    if gsi_key is None:
+        raise Exception("Table does not have a GSI")
+
+    # Make query params
+    PREFIX = "#dlp__"
+
+    # create update expression
+    key_condition_expression = f"{PREFIX}{gsi_key} = :{gsi_key}"
+
+    # create expression attribute values
+    expression_attribute_values = {f":{gsi_key}": gsi_value}
+
+    # create expression attribute names to prevent reserved words error
+    expression_attribute_names = {f"{PREFIX}{gsi_key}": gsi_key}
+
+    # initialize response items
+    response_items = []
+    table = dynamodb.Table(table_name)
+    has_more_items = True
+    last_evaluated_key = None
+
+    # Get items through pagination
+    while has_more_items:
+        query_params = {
+            "IndexName": gsi_key,
+            "KeyConditionExpression": key_condition_expression,
+            "ExpressionAttributeValues": expression_attribute_values,
+            "ExpressionAttributeNames": expression_attribute_names,
+        }
+        if last_evaluated_key is not None:
+            query_params["ExclusiveStartKey"] = last_evaluated_key
+
+        response = table.query(**query_params)
+        response_items.extend(response["Items"])
+        last_evaluated_key = response.get("LastEvaluatedKey")
+        has_more_items = last_evaluated_key is not None
+
+    return response_items
+
+
+def create_dynamo_item(table_name: str, input_item: dict) -> bool:
```

Comment:
@farisdurrani keep in mind that for any dynamodb table that will get created, the table must be written in terraform. do you think you can have functionality to automatically update the terraform config for the new table?

Reply:
or is that overengineering?

Reply:
I think that's overcomplicating. It is a reasonable assumption that when you're adding stuff to a table, the table must have existed already. Plus, there is a check to see if the table's details (e.g. name, id column) have already been registered in the app as a constant

Reply:
Got it. Fair point!